File: day17task/student.py
class Student:
    def __init__(self, name, age, department):
        if not isinstance(age, int) or age <= 0:
            raise ValueError(f"Age must be a positive integer, got {age}")
        self.name = name
        self.age = age
        self.department = department

    def __str__(self):
        return f"Student({self.name}, {self.age}, {self.department})"

    def display_details(self):
        print(f"Name: {self.name}")
        print(f"Age: {self.age}")
        print(f"Department: {self.department}")

    def calculate_grade(self, marks):
        if not (0 <= marks <= 100):
            raise ValueError(f"Marks must be between 0 and 100, got {marks}")
        thresholds = [(90, "A"), (80, "B"), (70, "C"), (60, "D")]
        grade = next((g for threshold, g in thresholds if marks >= threshold), "F")
        return grade


def calculate_factorial(n):
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    if n == 0 or n == 1:
        return 1
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result


# Example usage
if __name__ == "__main__":
    s = Student("Alice", 20, "Computer Science")
    s.display_details()
    grade = s.calculate_grade(85)
    print(f"{s.name}'s grade for 85 marks: {grade}")
    print(f"Factorial of 5: {calculate_factorial(5)}")

File: day17task/Student.java
public class Student {

    // Fields
    private String name;
    private int age;
    private String department;

    // Constructor
    public Student(String name, int age, String department) {
        this.name = name;
        this.age = age;
        this.department = department;
    }

    // Display student details
    public void displayDetails() {
        System.out.println("Name: " + name);
        System.out.println("Age: " + age);
        System.out.println("Department: " + department);
    }

    // Calculate grade based on marks
    public String calculateGrade(int marks) {
        String grade;
        if (marks >= 90) {
            grade = "A";
        } else if (marks >= 80) {
            grade = "B";
        } else if (marks >= 70) {
            grade = "C";
        } else if (marks >= 60) {
            grade = "D";
        } else {
            grade = "F";
        }
        System.out.println(name + "'s grade for " + marks + " marks: " + grade);
        return grade;
    }

    // Static method to calculate factorial
    public static long calculateFactorial(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Factorial is not defined for negative numbers");
        }
        if (n == 0 || n == 1) {
            return 1;
        }
        long result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }

    // Main method (entry point)
    public static void main(String[] args) {
        Student s = new Student("Alice", 20, "Computer Science");
        s.displayDetails();
        s.calculateGrade(85);

        // Factorial example
        System.out.println("Factorial of 5: " + calculateFactorial(5));
    }
}
